File: backend/app/schemas/club.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


# ── Club ──────────────────────────────────────────────────────────────

class ClubCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)


class ClubUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)


class ClubMemberOut(BaseModel):
    user_id: uuid.UUID
    username: str
    display_name: str | None
    avatar: str | None
    role: str
    joined_at: datetime

    model_config = {"from_attributes": True}


class ClubOut(BaseModel):
    id: uuid.UUID
    name: str
    description: str | None
    avatar: str | None
    owner_id: uuid.UUID
    member_count: int
    my_role: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class ClubDetailOut(ClubOut):
    members: list[ClubMemberOut]


# ── Invites ───────────────────────────────────────────────────────────

class InviteCreate(BaseModel):
    max_uses: int | None = Field(None, ge=1)
    expires_in_hours: int | None = Field(None, ge=1, le=720)


class InviteOut(BaseModel):
    id: uuid.UUID
    code: str
    url: str
    max_uses: int | None
    use_count: int
    expires_at: datetime | None
    active: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class JoinResult(BaseModel):
    club_id: uuid.UUID
    club_name: str
    role: str


# ── Leaderboard ───────────────────────────────────────────────────────

class LeaderboardEntry(BaseModel):
    user_id: uuid.UUID
    username: str
    display_name: str | None
    avatar: str | None
    best_score: int
    best_x_count: int
    session_id: uuid.UUID
    achieved_at: datetime


class LeaderboardOut(BaseModel):
    template_id: uuid.UUID
    template_name: str
    entries: list[LeaderboardEntry]


# ── Activity ──────────────────────────────────────────────────────────

class ActivityItem(BaseModel):
    type: str  # "session_completed" or "personal_record"
    user_id: uuid.UUID
    username: str
    display_name: str | None
    avatar: str | None
    template_name: str
    score: int
    x_count: int
    session_id: uuid.UUID
    occurred_at: datetime


# ── Events ────────────────────────────────────────────────────────────

class EventCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = Field(None, max_length=1000)
    template_id: uuid.UUID
    event_date: datetime
    location: str | None = Field(None, max_length=200)


class EventUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=200)
    description: str | None = Field(None, max_length=1000)
    event_date: datetime | None = None
    location: str | None = Field(None, max_length=200)


class EventParticipantOut(BaseModel):
    user_id: uuid.UUID
    username: str
    display_name: str | None
    avatar: str | None
    status: str
    score: int | None = None
    x_count: int | None = None
    session_id: uuid.UUID | None = None


class EventOut(BaseModel):
    id: uuid.UUID
    club_id: uuid.UUID
    name: str
    description: str | None
    template_id: uuid.UUID
    template_name: str | None = None
    event_date: datetime
    location: str | None
    created_by: uuid.UUID
    participants: list[EventParticipantOut]
    created_at: datetime

    model_config = {"from_attributes": True}


class EventRSVP(BaseModel):
    status: str = Field(pattern=r"^(going|maybe|declined)$")

File: mobile/lib/features/clubs/models/club.dart
class Club {
  final String id;
  final String name;
  final String? description;
  final String? avatar;
  final String ownerId;
  final int memberCount;
  final String? myRole;
  final DateTime createdAt;

  const Club({
    required this.id,
    required this.name,
    this.description,
    this.avatar,
    required this.ownerId,
    required this.memberCount,
    this.myRole,
    required this.createdAt,
  });

  factory Club.fromJson(Map<String, dynamic> json) {
    return Club(
      id: json['id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      avatar: json['avatar'] as String?,
      ownerId: json['owner_id'] as String,
      memberCount: json['member_count'] as int? ?? 0,
      myRole: json['my_role'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

class ClubDetail {
  final String id;
  final String name;
  final String? description;
  final String? avatar;
  final String ownerId;
  final int memberCount;
  final String? myRole;
  final DateTime createdAt;
  final List<ClubMember> members;

  const ClubDetail({
    required this.id,
    required this.name,
    this.description,
    this.avatar,
    required this.ownerId,
    required this.memberCount,
    this.myRole,
    required this.createdAt,
    this.members = const [],
  });

  factory ClubDetail.fromJson(Map<String, dynamic> json) {
    return ClubDetail(
      id: json['id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      avatar: json['avatar'] as String?,
      ownerId: json['owner_id'] as String,
      memberCount: json['member_count'] as int? ?? 0,
      myRole: json['my_role'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      members: (json['members'] as List?)
              ?.map((m) => ClubMember.fromJson(m as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

class ClubMember {
  final String userId;
  final String username;
  final String? displayName;
  final String? avatar;
  final String role;
  final DateTime joinedAt;

  const ClubMember({
    required this.userId,
    required this.username,
    this.displayName,
    this.avatar,
    required this.role,
    required this.joinedAt,
  });

  factory ClubMember.fromJson(Map<String, dynamic> json) {
    return ClubMember(
      userId: json['user_id'] as String,
      username: json['username'] as String,
      displayName: json['display_name'] as String?,
      avatar: json['avatar'] as String?,
      role: json['role'] as String,
      joinedAt: DateTime.parse(json['joined_at'] as String),
    );
  }

  String get effectiveName => displayName ?? username;
}

class LeaderboardGroup {
  final String templateId;
  final String templateName;
  final List<LeaderboardEntry> entries;

  const LeaderboardGroup({
    required this.templateId,
    required this.templateName,
    this.entries = const [],
  });

  factory LeaderboardGroup.fromJson(Map<String, dynamic> json) {
    return LeaderboardGroup(
      templateId: json['template_id'] as String,
      templateName: json['template_name'] as String,
      entries: (json['entries'] as List?)
              ?.map((e) => LeaderboardEntry.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

class LeaderboardEntry {
  final String userId;
  final String username;
  final String? displayName;
  final String? avatar;
  final int bestScore;
  final int bestXCount;
  final String sessionId;
  final DateTime achievedAt;

  const LeaderboardEntry({
    required this.userId,
    required this.username,
    this.displayName,
    this.avatar,
    required this.bestScore,
    required this.bestXCount,
    required this.sessionId,
    required this.achievedAt,
  });

  factory LeaderboardEntry.fromJson(Map<String, dynamic> json) {
    return LeaderboardEntry(
      userId: json['user_id'] as String,
      username: json['username'] as String,
      displayName: json['display_name'] as String?,
      avatar: json['avatar'] as String?,
      bestScore: json['best_score'] as int? ?? 0,
      bestXCount: json['best_x_count'] as int? ?? 0,
      sessionId: json['session_id'] as String,
      achievedAt: DateTime.parse(json['achieved_at'] as String),
    );
  }

  String get effectiveName => displayName ?? username;
}

class ActivityItem {
  final String type;
  final String userId;
  final String username;
  final String? displayName;
  final String? avatar;
  final String templateName;
  final int score;
  final int xCount;
  final String sessionId;
  final DateTime occurredAt;

  const ActivityItem({
    required this.type,
    required this.userId,
    required this.username,
    this.displayName,
    this.avatar,
    required this.templateName,
    required this.score,
    required this.xCount,
    required this.sessionId,
    required this.occurredAt,
  });

  factory ActivityItem.fromJson(Map<String, dynamic> json) {
    return ActivityItem(
      type: json['type'] as String,
      userId: json['user_id'] as String,
      username: json['username'] as String,
      displayName: json['display_name'] as String?,
      avatar: json['avatar'] as String?,
      templateName: json['template_name'] as String,
      score: json['score'] as int? ?? 0,
      xCount: json['x_count'] as int? ?? 0,
      sessionId: json['session_id'] as String,
      occurredAt: DateTime.parse(json['occurred_at'] as String),
    );
  }

  String get effectiveName => displayName ?? username;
}
